### PR TITLE
chore: Graceful shutdown for API Server (redo) (#18642)

### LINKED ISSUE #18642: Implement graceful shutdown in all Argo CD components
Procfile sets GOCOVDIR for five Argo CD components. But only one (applicationset-controller) shuts down gracefully enough for code coverage data to be produced. In other words, go writes a covmeta file, but not a covcounters file (the file that actually contains coverage info). So today we get partial code coverage info for e2e tests.

The task is to do two things for the remaining four components: 1) implement graceful shutdown and 2) [add the component's gocov directory to the e2e reporting CI](https://github.com/argoproj/argo-cd/blob/master/.github/workflows/ci-build.yaml#L367-L371).

- [x] applicationset-controller
- [x] controller
- [ ] api-server ([WIP](https://github.com/argoproj/argo-cd/pull/15938))
- [x] repo-server
- [x] notification

Exactly what "graceful shutdown means" may vary by component. Concretely, I think it means we need to listen to shutdown signals and end all running goroutines. But the "graceful" part depends on what's running.

### PR DESCRIPTION
**Pull request by andrii-korotkov-verkada**
Closes #18642
    
Implements a graceful shutdown the the API server. Without this, ArgoCD API server will eventually return 502 during rolling update. However, healthcheck would return 503 if the server is terminating.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -712,8 +723,8 @@ func (a *ArgoCDServer) Run(ctx context.Context, listeners *Listeners) {
 	select {
 	case signal := <-a.stopCh:
 		log.Infof("API Server received signal: %s", signal.String())
-		// SIGUSR1 is used for triggering a server restart
-		if signal != syscall.SIGUSR1 {
+		gracefulRestartSignal := GracefulRestartSignal{}
+		if signal != gracefulRestartSignal {
```
Comment (leoluz) [inline server/server.go:727]: minor but do you need to allocate the variable to do the comparison or just doing `signal != gracefulRestartSignal{}` would work?

**Review by leoluz — COMMENTED** [review]
Tks for the fix.
Added a minor comment.

File: server/server.go (hunk 1 above)
Comment (andrii-korotkov-verkada) [reply server/server.go:727]: I've tried that, but for some reason VSCode would highlight that with red.

Comment (andrii-korotkov-verkada) [reply server/server.go:727]: Was telling that ; should be there not {

**Review by leoluz — APPROVED** [review]
LGTM

**Pull request merged**